File: frontend/src/lib/api/franchise.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiClient } from './client'

// ── Types ──────────────────────────────────────────────────────────────────

export interface Franchisee {
  id: string
  name: string
  branch_name: string
  location: string
  contact: string
  status: 'active' | 'inactive' | 'terminated'
  created_by: string
  user_id?: string
  created_at: string
  updated_at: string
}

export interface FranchiseAgreement {
  id: string
  franchisee_id: string
  buy_in_fee: number
  monthly_royalty: number
  revenue_royalty_pct: number
  start_date: string
  end_date?: string
  status: 'active' | 'inactive' | 'terminated'
  created_at: string
  updated_at: string
}

export interface RoyaltyRecord {
  id: string
  franchise_agreement_id: string
  period: string
  gross_revenue: number
  monthly_royalty: number
  revenue_royalty: number
  total_royalty: number
  status: 'unpaid' | 'overdue' | 'paid'
  created_at: string
  paid_at?: string
  recorded_by: string
}

// ── Hooks ──────────────────────────────────────────────────────────────────

export function useMyFranchisee() {
  return useQuery({
    queryKey: ['franchisee', 'me'],
    queryFn: () => apiClient.get<Franchisee>('/me/franchisee').then((r) => r.data),
  })
}

export function useFranchisee(id: string) {
  return useQuery({
    queryKey: ['franchisee', id],
    queryFn: () => apiClient.get<Franchisee>(`/franchisees/${id}`).then((r) => r.data),
    enabled: !!id,
  })
}

export function useAgreement(franchiseeId: string) {
  return useQuery({
    queryKey: ['franchise-agreement', franchiseeId],
    queryFn: () =>
      apiClient
        .get<FranchiseAgreement>(`/franchise-agreements/${franchiseeId}`)
        .then((r) => r.data),
    enabled: !!franchiseeId,
  })
}

export function useRoyaltyRecords(franchiseeId: string) {
  return useQuery({
    queryKey: ['royalty-records', franchiseeId],
    queryFn: () =>
      apiClient
        .get<RoyaltyRecord[]>(`/royalty-records/${franchiseeId}/all`)
        .then((r) => r.data),
    enabled: !!franchiseeId,
  })
}

export function useMarkRoyaltyPaid() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (id: string) =>
      apiClient.post(`/royalty-records/${id}/mark-paid`).then((r) => r.data),
    onSuccess: () => qc.invalidateQueries({ queryKey: ['royalty-records'] }),
  })
}

File: api/internal/domain/franchise/franchise.go
package franchise

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
)

var (
	ErrFranchiseeNotFound    = errors.New("franchisee not found")
	ErrAgreementNotFound     = errors.New("franchise agreement not found")
	ErrRoyaltyRecordNotFound = errors.New("royalty payment record not found")
	ErrOtherRevenueNotFound  = errors.New("other revenue not found")
)

type Franchisee struct {
	ID                 uuid.UUID
	Name               string
	BranchName         string
	Location           string
	Contact            string
	Status             string
	CreatedBy          *uuid.UUID
	CreatedAt          time.Time
	UpdatedAt          time.Time
	AgreementStartDate string
	AgreementEndDate   string
}

type FranchiseAgreement struct {
	ID                uuid.UUID
	FranchiseeID      uuid.UUID
	BuyInFee          float64
	MonthlyRoyalty    float64
	RevenueRoyaltyPct float64
	StartDate         string
	EndDate           string
	Status            string
	CreatedAt         time.Time
	UpdatedAt         time.Time
}

type RoyaltyPaymentRecord struct {
	ID                   uuid.UUID
	FranchiseAgreementID uuid.UUID
	Period              string
	GrossRevenue        float64
	MonthlyRoyalty      float64
	RevenueRoyalty      float64
	TotalRoyalty        float64
	Status              string
	PaidAt              *time.Time
	RecordedBy          *uuid.UUID
	CreatedAt           time.Time
	UpdatedAt           time.Time
}

type BranchOtherRevenue struct {
	ID           uuid.UUID
	FranchiseeID uuid.UUID
	Label        string
	Amount       float64
	RevenueDate  string
	AddedBy      *uuid.UUID
	CreatedAt    time.Time
	UpdatedAt    time.Time
}

type WriteRepository interface {
	SaveFranchisee(ctx context.Context, f *Franchisee) error
	UpdateFranchisee(ctx context.Context, f *Franchisee) error
	SaveAgreement(ctx context.Context, a *FranchiseAgreement) error
	UpdateAgreement(ctx context.Context, a *FranchiseAgreement) error
	SaveRoyaltyPayment(ctx context.Context, r *RoyaltyPaymentRecord) error
	MarkRoyaltyPaid(ctx context.Context, id uuid.UUID, paidAt time.Time) error
	SaveOtherRevenue(ctx context.Context, r *BranchOtherRevenue) error
	UpdateOtherRevenue(ctx context.Context, r *BranchOtherRevenue) error
	DeleteOtherRevenue(ctx context.Context, id uuid.UUID) error
}

type ReadRepository interface {
	GetFranchiseeByID(ctx context.Context, id uuid.UUID) (*Franchisee, error)
	ListFranchisees(ctx context.Context, offset, limit int, status, search string) ([]*Franchisee, int, error)
	GetAgreementByFranchiseeID(ctx context.Context, franchiseeID uuid.UUID) (*FranchiseAgreement, error)
	GetAgreementByID(ctx context.Context, id uuid.UUID) (*FranchiseAgreement, error)
	ListRoyaltyPayments(ctx context.Context, franchiseeID uuid.UUID, period string) ([]*RoyaltyPaymentRecord, error)
	GetRoyaltyPaymentByID(ctx context.Context, id uuid.UUID) (*RoyaltyPaymentRecord, error)
	ListOtherRevenues(ctx context.Context, franchiseeID uuid.UUID, period string) ([]*BranchOtherRevenue, error)
	GetOtherRevenueByID(ctx context.Context, id uuid.UUID) (*BranchOtherRevenue, error)
}
